[LT] Upstream more util functions

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69277
* __->__ #69098

Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

desertfire committed:
[LT] Merge more util functions

Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

[ghstack-poisoned]

alanwaketan commented on torch/csrc/lazy/backend/backend_interface.cpp:
```diff
@@ -0,0 +1,41 @@
+namespace torch {
+namespace lazy {
+
+std::atomic<const BackendImplInterface*> backend_impl_registry;
```
Oops, you might need to rebase after I land: https://github.com/pytorch/pytorch/pull/69064.

desertfire replied:
Yes, I noticed that. I will merge after you land yours. I also realized I need to add TORCH_API in those header files.

alanwaketan replied:
Yeah, most of those are not inline.

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32758480).

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32758480).

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32758480).

alanwaketan commented on torch/csrc/lazy/core/helpers.h:
```diff
@@ -0,0 +1,70 @@
+#pragma once
```
Sounds like we could consolidate the file and util.h. Can you leave a TODO as a follow up?

desertfire replied:
Agree. I have mentioned it in the PR description, but let me add a TODO to make it explicit.

alanwaketan commented on torch/csrc/lazy/core/tensor_util.h:
```diff
@@ -0,0 +1,68 @@
+    const std::vector<BackendDevice>& devices);
+
+// Makes a deep copy of an ATEN tensor.
+static inline at::Tensor CopyTensor(const at::Tensor& ref) {
```
I don't think you need `static inline` in functions defined in headers. Is it coding style?

desertfire replied:
I have seen both `static inline` and `inline` get used in core. There is actually a lint warning for `static`, so this might be a good time to remove it.

alanwaketan replied:
Just for my education, what are the benefits/effects of using static in non-member functions in headers? For `inline`, functions defined in header should be implicitly `inline` unless I missed something.

alanwaketan commented on torch/csrc/lazy/core/tensor_util.h:
```diff
@@ -0,0 +1,68 @@
+}
+
+// Same as above, with an additional cast.
+static inline at::Tensor CopyTensor(
```
I don't think you need static inline in functions defined in headers. Is it coding style?

alanwaketan commented on torch/csrc/lazy/core/tensor_util.h:
```diff
@@ -0,0 +1,68 @@
+}
+
+// Unwraps tensor to target dtype if it's a wrapped number.
+static inline at::Tensor UnwrapNumber(
```
I don't think you need static inline in functions defined in headers. Is it coding style?

alanwaketan reviewed (approved):
LGTM in general. Can you please address some of the comments before landing?

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32758480).

desertfire committed:
Update on "[LT] Upstream more util functions"


Summary: Add the following utils: helpers, ir_dump_util, and
tensor_util. Some of the util functions may be better organized by
grouping into different files, but we can leave that for later.

Differential Revision: [D32758480](https://our.internmc.facebook.com/intern/diff/D32758480)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32758480).
